src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -486,18 +494,22 @@ impl<O: ForestObligation> ObligationForest<O> {
 
         for node in &self.nodes {
             if node.state.get() == NodeState::Pending {
-                self.mark_as_waiting_from(node)
+                if let Some(parent) = node.parent {
+                    self.mark_as_waiting_from(&self.nodes[parent.get()]);
+                }
+
+                for dependent in &node.dependents {
+                    self.mark_as_waiting_from(&self.nodes[dependent.get()]);
+                }
```

Comment:
rather than making two copies of this code, maybe pull out a helper like `mark_neighbors_as_waiting_from(node)`?